p2p/service.py

```diff
@@ -31,7 +35,12 @@ def __init__(self, token: CancelToken=None) -> None:
         self.cleaned_up = asyncio.Event()
         self._child_services = []
 
-        base_token = CancelToken(type(self).__name__)
+        if loop is None:
+            base_token = CancelToken(type(self).__name__)
+        else:
+            self.target_loop = loop
+            base_token = CancelToken(type(self).__name__, loop=loop)
```

Comment:
I think you can drop this `if/else`.  Passing in `loop=None` to `CancelToken` should be equivalent to not including the argument.